resolwe_bio/processes/clustering/hierarchical_clustering.py

```diff
@@ -0,0 +1,899 @@
+                    ("Rattus norvegicus", "Rattus norvegicus"),
+                    ("Dictyostelium discoideum", "Dictyostelium discoideum"),
+                ],
+            )
+            log2 = BooleanField(
+                label="Log-transform expressions",
+                default=True,
+                description="Transform expressions with log2(x + 1) before clustering.",
+            )
+            z_score = BooleanField(
+                label="Z-score normalization",
+                default=True,
+                description="Use Z-score normalization of gene expressions before clustering.",
+            )
+
+        class Processing:
+            """Processing."""
+
+            distance_metric = StringField(
+                label="Distance metric",
+                default="pearson",
+                choices=[
+                    ("euclidean", "Euclidean"),
+                    ("pearson", "Pearson"),
+                    ("spearman", "Spearman"),
+                ],
+            )
+            linkage_method = StringField(
+                label="Linkage method",
+                default="average",
+                choices=[
+                    ("single", "single"),
+                    ("average", "average"),
+                    ("complete", "complete"),
+                ],
+            )
+
+        class Postprocessing:
+            """Postprocessing."""
+
+            order = BooleanField(
+                label="Order samples optimally",
+                default=True,
+            )
+
+        preprocessing = GroupField(
+            Preprocessing, label="Preprocessing", hidden="!advanced"
+        )
+        processing = GroupField(Processing, label="Processing", hidden="!advanced")
+        postprocessing = GroupField(
+            Postprocessing, label="Postprocessing", hidden="!advanced"
+        )
+
+    class Output:
+        """Output fields to process HierarchicalClusteringSamples."""
+
+        cluster = JsonField(
+            label="Hierarchical clustering",
+            required=False,
+        )
+
+    def get_clustering_samples(
+        self,
+        expressions,
+        distance_metric="euclidean",
+        linkage_method="average",
+        order=False,
+    ):
+        """Compute linkage, order, and produce a dendrogram."""
+        try:
+            link = linkage(
+                y=expressions.transpose(),
+                method=linkage_method,
+                metric=distance_metric,
+                optimal_ordering=order,
+            )
+        except Exception:
+            self.error("Cannot compute linkage.")
+        try:
+            dend = dendrogram(link, no_plot=True)
+        except Exception:
+            self.error("Cannot compute dendrogram.")
+        return link, dend
+
+    def run(self, inputs, outputs):
+        """Run analysis."""
+
+        sample_files = []
+        sample_ids = []
+        sample_names = []
+        gene_labels = []
+
+        for exp in inputs.exps:
```

Comment:
The code for `source`, `species` and `exp_type` checks is duplicated between the two processes in the same file. This checks could be written as functions and reused in two places.